compiler/rustc_middle/src/hooks/mod.rs

```diff
@@ -0,0 +1,65 @@
+use crate::mir;
+use crate::query::TyCtxtAt;
+use crate::ty::{Ty, TyCtxt};
+use rustc_span::DUMMY_SP;
+
+macro_rules! declare_hooks {
```

Comment:
Would be good to have some docs in this file that explain hat a "hook" is and how it differs from a "query".